src/backend/vulkan/src/lib.rs

```diff
@@ -680,6 +735,152 @@ pub struct DeviceCreationFeatures {
     imageless_framebuffers: Option<vk::PhysicalDeviceImagelessFramebufferFeaturesKHR>,
 }
 
+/// A declaration of an extension and its dependencies.
+#[derive(Debug)]
+struct Extension {
+    name: &'static CStr,
+    required_version: Version,
+    dependencies: Vec<&'static CStr>,
+    promoted_version: Option<Version>,
+}
+
+impl Extension {
+    pub fn new(name: &'static CStr, required_version: Version) -> Self {
+        Self {
+            name,
+            required_version,
+            dependencies: Vec::new(),
+            promoted_version: None,
+        }
+    }
+
+    pub fn with_dependencies(mut self, dependencies: Vec<&'static CStr>) -> Self {
+        self.dependencies = dependencies;
+        self
+    }
+
+    pub fn promoted_to(mut self, promoted_version: Version) -> Self {
+        self.promoted_version = Some(promoted_version);
+        self
+    }
+
+    pub fn is_compatible_with_instance_version(&self, instance_version: Version) -> bool {
+        self.required_version.0
+            <= vk::make_version(instance_version.major(), instance_version.minor(), 0)
+    }
+
+    pub fn is_promoted_by_instance_version(&self, instance_version: Version) -> bool {
+        if let Some(promoted_version) = self.promoted_version {
+            vk::make_version(instance_version.major(), instance_version.minor(), 0)
+                >= promoted_version.0
+        } else {
+            false
+        }
+    }
+}
+
+/// Container to keep track of the extensions we use and their dependencies. Used for `Self::resolve_dependencies`.
+#[derive(Debug)]
+struct ExtensionsRegistry {
+    extensions: HashMap<&'static CStr, Extension>,
+}
+
+impl ExtensionsRegistry {
+    pub fn new<I>(extensions: I) -> Self
+    where
+        I: IntoIterator<Item = Extension>,
+    {
+        Self {
+            extensions: extensions
+                .into_iter()
+                .map(|extension| (extension.name, extension))
+                .collect(),
+        }
+    }
+
+    /// Resolve `requested_extensions` into the full list of transitive dependencies and filter out any extensions that are not supported or are no longer explicitly needed.
+    ///
+    /// In general, `requested_extensions` should include the extension list needed to support the oldest version of Vulkan.
+    // TODO: Should we signal if an extension that was request could not be added?
```

Comment:
yeah, looks like we need to define the interface here a bit stronger. Does this function guarantee the resolve? or are some of the requested extensions optional?

Reply:
I'll make it return a `Result` that indicates an expected extension didn't make it. In `PhysicalDevice::open` it'll get mapped to `DeviceCreationError::MissingExtension`.